{"payload":{"allShortcutsEnabled":false,"fileTree":{"":{"items":[{"name":"Conserved_PCGs_db","path":"Conserved_PCGs_db","contentType":"directory"},{"name":"bin","path":"bin","contentType":"directory"},{"name":"container","path":"container","contentType":"directory"},{"name":"misc","path":"misc","contentType":"directory"},{"name":"modules","path":"modules","contentType":"directory"},{"name":"README.md","path":"README.md","contentType":"file"}],"totalCount":6}},"fileTreeProcessingTime":2.4721119999999996,"foldersToFetch":[],"reducedMotionEnabled":null,"repo":{"id":636572105,"defaultBranch":"main","name":"PMAT","ownerLogin":"bichangwei","currentUserCanPush":false,"isFork":false,"isEmpty":false,"createdAt":"2023-05-05T06:45:18.000Z","ownerAvatar":"https://avatars.githubusercontent.com/u/19208952?v=4","public":true,"private":false,"isOrgOwned":false},"refInfo":{"name":"main","listCacheKey":"v0:1690464270.0","canEdit":false,"refType":"branch","currentOid":"f536ea3ebe8563418e096e3fc2b325e5fbf4ba7b"},"path":"README.md","currentUser":null,"blob":{"rawLines":null,"stylingDirectives":null,"csv":null,"csvError":null,"dependabotInfo":{"showConfigurationBanner":false,"configFilePath":null,"networkDependabotPath":"/bichangwei/PMAT/network/updates","dismissConfigurationNoticePath":"/settings/dismiss-notice/dependabot_configuration_notice","configurationNoticeDismissed":null,"repoAlertsPath":"/bichangwei/PMAT/security/dependabot","repoSecurityAndAnalysisPath":"/bichangwei/PMAT/settings/security_analysis","repoOwnerIsOrg":false,"currentUserCanAdminRepo":false},"displayName":"README.md","displayUrl":"https://github.com/bichangwei/PMAT/blob/main/README.md?raw=true","headerInfo":{"blobSize":"4.28 KB","deleteInfo":{"deletePath":null,"deleteTooltip":"You must be signed in to make or propose changes"},"editInfo":{"editTooltip":"You must be signed in to make or propose changes"},"ghDesktopPath":"https://desktop.github.com","gitLfsPath":null,"onBranch":true,"shortPath":"3186b4e","siteNavLoginPath":"/login?return_to=https%3A%2F%2Fgithub.com%2Fbichangwei%2FPMAT%2Fblob%2Fmain%2FREADME.md","isCSV":false,"isRichtext":true,"toc":[{"level":1,"text":"PMAT - an efficient assembly toolkit for plant mitochondrial genome","anchor":"pmat---an-efficient-assembly-toolkit-for-plant-mitochondrial-genome","htmlText":"PMAT - an efficient assembly toolkit for plant mitochondrial genome"},{"level":2,"text":"Installation","anchor":"installation","htmlText":"Installation"},{"level":4,"text":"recommended installation method","anchor":"recommended-installation-method","htmlText":"recommended installation method"},{"level":2,"text":"Repuirement","anchor":"repuirement","htmlText":"Repuirement"},{"level":6,"text":"When inputting HiFi reads, the following software are required:","anchor":"when-inputting-hifi-reads-the-following-software-are-required","htmlText":"When inputting HiFi reads, the following software are required:"},{"level":6,"text":"When inputting CLR or ONT reads, the following software are required:","anchor":"when-inputting-clr-or-ont-reads-the-following-software-are-required","htmlText":"When inputting CLR or ONT reads, the following software are required:"},{"level":2,"text":"options and usage","anchor":"options-and-usage","htmlText":"options and usage"},{"level":2,"text":"Tips","anchor":"tips","htmlText":"Tips"},{"level":2,"text":"Version","anchor":"version","htmlText":"Version"},{"level":2,"text":"Author","anchor":"author","htmlText":"Author"}],"lineInfo":{"truncatedLoc":"103","truncatedSloc":"72"},"mode":"file"},"image":false,"isCodeownersFile":null,"isValidLegacyIssueTemplate":false,"issueTemplateHelpUrl":"https://docs.github.com/articles/about-issue-and-pull-request-templates","issueTemplate":null,"discussionTemplate":null,"language":"Markdown","large":false,"loggedIn":false,"newDiscussionPath":"/bichangwei/PMAT/discussions/new","newIssuePath":"/bichangwei/PMAT/issues/new","planSupportInfo":{"repoIsFork":null,"repoOwnedByCurrentUser":null,"requestFullPath":"/bichangwei/PMAT/blob/main/README.md","showFreeOrgGatedFeatureMessage":null,"showPlanSupportBanner":null,"upgradeDataAttributes":null,"upgradePath":null},"publishBannersInfo":{"dismissActionNoticePath":"/settings/dismiss-notice/publish_action_from_dockerfile","dismissStackNoticePath":"/settings/dismiss-notice/publish_stack_from_file","releasePath":"/bichangwei/PMAT/releases/new?marketplace=true","showPublishActionBanner":false,"showPublishStackBanner":false},"renderImageOrRaw":false,"richText":"<article class=\"markdown-body entry-content container-lg\" itemprop=\"text\"><h1 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-pmat---an-efficient-assembly-toolkit-for-plant-mitochondrial-genome\" class=\"anchor\" aria-hidden=\"true\" href=\"#pmat---an-efficient-assembly-toolkit-for-plant-mitochondrial-genome\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>PMAT - an efficient assembly toolkit for plant mitochondrial genome</h1>\n<p align=\"center\" dir=\"auto\"><a target=\"_blank\" rel=\"noopener noreferrer\" href=\"/bichangwei/PMAT/blob/main/misc/logo.png\"><img src=\"/bichangwei/PMAT/raw/main/misc/logo.png\" alt=\"PMAT\" width=\"600\" style=\"max-width: 100%;\"></a></p>\n<p dir=\"auto\">PMAT is an efficient toolkit for mitochondrial genome assembly. It can correct three-generation sequencing data using <a href=\"https://github.com/marbl/canu\">canu</a> or <a href=\"https://github.com/Nextomics/NextDenovo\">NextDenovo</a> and then use <a href=\"https://evomics.org/learning/assembly-and-alignment/newbler/\" rel=\"nofollow\">Newbler</a> to assemble it. Finally, PMAT can identify mitochondrial genome sequences and generate mitochondrial genome structures.</p>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-installation\" class=\"anchor\" aria-hidden=\"true\" href=\"#installation\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>Installation</h2>\n<h4 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-recommended-installation-method\" class=\"anchor\" aria-hidden=\"true\" href=\"#recommended-installation-method\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>recommended installation method</h4>\n<div class=\"highlight highlight-source-shell notranslate position-relative overflow-auto\" dir=\"auto\" data-snippet-clipboard-copy-content=\"    wget https://github.com/bichangwei/PMAT/releases/download/v1.1.0/PMAT-1.1.0.tar.gz\n    tar -zxvf PMAT-1.1.0.tar.gz\n    cd PMAT/bin\n    chmod a+x PMAT\n    PMAT --help\"><pre>    wget https://github.com/bichangwei/PMAT/releases/download/v1.1.0/PMAT-1.1.0.tar.gz\n    tar -zxvf PMAT-1.1.0.tar.gz\n    <span class=\"pl-c1\">cd</span> PMAT/bin\n    chmod a+x PMAT\n    PMAT --help</pre></div>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-repuirement\" class=\"anchor\" aria-hidden=\"true\" href=\"#repuirement\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>Repuirement</h2>\n<h6 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-when-inputting-hifi-reads-the-following-software-are-required\" class=\"anchor\" aria-hidden=\"true\" href=\"#when-inputting-hifi-reads-the-following-software-are-required\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>When inputting HiFi reads, the following software are required:</h6>\n<ul dir=\"auto\">\n<li><a href=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download\" rel=\"nofollow\"><strong>blastn</strong></a>  Needs to be installed in the <code>PATH</code>.</li>\n<li><a href=\"https://github.com/YanshuQu/runAssembly\"><strong>singularity</strong></a></li>\n</ul>\n<h6 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-when-inputting-clr-or-ont-reads-the-following-software-are-required\" class=\"anchor\" aria-hidden=\"true\" href=\"#when-inputting-clr-or-ont-reads-the-following-software-are-required\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>When inputting CLR or ONT reads, the following software are required:</h6>\n<ul dir=\"auto\">\n<li><a href=\"https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download\" rel=\"nofollow\"><strong>blastn</strong></a>     Needs to be installed in the <code>PATH</code>.</li>\n<li><a href=\"https://github.com/marbl/canu\"><strong>canu</strong></a> or <a href=\"https://github.com/Nextomics/NextDenovo\"><strong>NextDenovo</strong></a>    Suggested to be installed in the <code>PATH</code>.</li>\n<li><a href=\"https://github.com/YanshuQu/runAssembly\"><strong>singularity</strong></a></li>\n</ul>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-options-and-usage\" class=\"anchor\" aria-hidden=\"true\" href=\"#options-and-usage\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>options and usage</h2>\n<p dir=\"auto\">Run <code>PMAT --help</code> to view the program's usage guide.</p>\n<div class=\"snippet-clipboard-content notranslate position-relative overflow-auto\" data-snippet-clipboard-copy-content=\"    usage: PMAT &lt;command&gt; &lt;arguments&gt;\n    \n     ______     ___           __        ____       _____________ \n    |   __  \\  |   \\        /   |      / __ \\     |_____   _____|\n    |  |__)  | | |\\ \\      / /| |     / /  \\ \\          | |      \n    |   ____/  | | \\ \\    / / | |    / /____\\ \\         | |      \n    |  |       | |  \\ \\  / /  | |   / /______\\ \\        | |      \n    |  |       | |   \\ \\/ /   | |  / /        \\ \\       | |      \n    |__|       |_|    \\__/    |_| /_/          \\_\\      |_|\n\n\n    PMAT            an efficient assembly toolkit for plant mitochondrial genome\n    Version         1.1.0\n    Contributors    Bi,C. and Han,F.\n    Email           bichwei@njfu.edu.cn, hanfc@caf.ac.cn\n    \n    For more information about PMAT, see https://github.com/bichangwei/PMAT\n\n    optional arguments:\n        -h, --help     show this help message and exit\n        -v, --version  show program's version number and exit\n\n    Commands:\n\n        autoMito         an efficient assembly toolkit for plant mitochondrial genome\n    \n        graphBuild  Structure assembly based on Newbler output\"><pre class=\"notranslate\"><code>    usage: PMAT &lt;command&gt; &lt;arguments&gt;\n    \n     ______     ___           __        ____       _____________ \n    |   __  \\  |   \\        /   |      / __ \\     |_____   _____|\n    |  |__)  | | |\\ \\      / /| |     / /  \\ \\          | |      \n    |   ____/  | | \\ \\    / / | |    / /____\\ \\         | |      \n    |  |       | |  \\ \\  / /  | |   / /______\\ \\        | |      \n    |  |       | |   \\ \\/ /   | |  / /        \\ \\       | |      \n    |__|       |_|    \\__/    |_| /_/          \\_\\      |_|\n\n\n    PMAT            an efficient assembly toolkit for plant mitochondrial genome\n    Version         1.1.0\n    Contributors    Bi,C. and Han,F.\n    Email           bichwei@njfu.edu.cn, hanfc@caf.ac.cn\n    \n    For more information about PMAT, see https://github.com/bichangwei/PMAT\n\n    optional arguments:\n        -h, --help     show this help message and exit\n        -v, --version  show program's version number and exit\n\n    Commands:\n\n        autoMito         an efficient assembly toolkit for plant mitochondrial genome\n    \n        graphBuild  Structure assembly based on Newbler output\n</code></pre></div>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-tips\" class=\"anchor\" aria-hidden=\"true\" href=\"#tips\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>Tips</h2>\n<ul dir=\"auto\">\n<li>\n<p dir=\"auto\"><strong>autoMito</strong> -- De novo analysis of sequencing data (hifi, clr and ont).</p>\n<p dir=\"auto\">run <code>PMAT autoMito --help</code> view the usage guide for the autoMito command.</p>\n<p dir=\"auto\">Example:</p>\n<div class=\"snippet-clipboard-content notranslate position-relative overflow-auto\" data-snippet-clipboard-copy-content=\"# If canu or NextDenovo is not installed in the PATH when using clr or ont data, you need to provide the parameter -cp or -np.\nFor HiFi data    : PMAT autoMito -i hifi.fastq.gz -o hifi_assembly -st hifi -g 500M\nFor ONT raw data : PMAT autoMito -i ont.fastq.gz -o ont_assembly -st ont -cfg ont_correct.cfg -tk autoMito -g 500M\nFor CLR raw data : PMAT autoMito -i ont.fastq.gz -o clr_assembly -st clr -cfg clr_correct.cfg -tk all -g 500M\nFor ONT corrected data : PMAT autoMito -i corrected.fa -o ont_assembly -st ont -tk p1 -g 500M\nFor CLR corrected data : PMAT autoMito -i corrected.fa -o clr_assembly -st clr -tk p1 -g 500M\"><pre class=\"notranslate\"><code># If canu or NextDenovo is not installed in the PATH when using clr or ont data, you need to provide the parameter -cp or -np.\nFor HiFi data    : PMAT autoMito -i hifi.fastq.gz -o hifi_assembly -st hifi -g 500M\nFor ONT raw data : PMAT autoMito -i ont.fastq.gz -o ont_assembly -st ont -cfg ont_correct.cfg -tk autoMito -g 500M\nFor CLR raw data : PMAT autoMito -i ont.fastq.gz -o clr_assembly -st clr -cfg clr_correct.cfg -tk all -g 500M\nFor ONT corrected data : PMAT autoMito -i corrected.fa -o ont_assembly -st ont -tk p1 -g 500M\nFor CLR corrected data : PMAT autoMito -i corrected.fa -o clr_assembly -st clr -tk p1 -g 500M\n</code></pre></div>\n<p dir=\"auto\">notes:</p>\n<div class=\"snippet-clipboard-content notranslate position-relative overflow-auto\" data-snippet-clipboard-copy-content=\"1. Make sure blastn is installed in the PATH.\n2. If canu or NextDenovo is not installed in the PATH when using clr or ont data, you need to provide the parameter -cp or -np.\"><pre class=\"notranslate\"><code>1. Make sure blastn is installed in the PATH.\n2. If canu or NextDenovo is not installed in the PATH when using clr or ont data, you need to provide the parameter -cp or -np.\n</code></pre></div>\n</li>\n<li>\n<p dir=\"auto\"><strong>graphBuild</strong> -- Structure assembly based on Newbler results.\nrun <code>PMAT graphBuild --help</code> view the usage guide for the graphBuild command.</p>\n<p dir=\"auto\">Example:</p>\n<div class=\"highlight highlight-source-shell notranslate position-relative overflow-auto\" dir=\"auto\" data-snippet-clipboard-copy-content=\"PMAT graphBuild -c PMATContigGraph.txt -a PMATAllContigs.fna -gs 500M -rs 4G -o output\nPMAT graphBuild -c PMATContigGraph.txt -a PMATAllContigs.fna -gs 500M -rs hifi.cut20k.fa -s 1 6 8 -o output\"><pre>PMAT graphBuild -c PMATContigGraph.txt -a PMATAllContigs.fna -gs 500M -rs 4G -o output\nPMAT graphBuild -c PMATContigGraph.txt -a PMATAllContigs.fna -gs 500M -rs hifi.cut20k.fa -s 1 6 8 -o output</pre></div>\n<p dir=\"auto\">notes:</p>\n<div class=\"snippet-clipboard-content notranslate position-relative overflow-auto\" data-snippet-clipboard-copy-content=\"  When using the graphBuild command, the blastn is required and needs to be installed in the PATH.\"><pre class=\"notranslate\"><code>  When using the graphBuild command, the blastn is required and needs to be installed in the PATH.\n</code></pre></div>\n</li>\n</ul>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-version\" class=\"anchor\" aria-hidden=\"true\" href=\"#version\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>Version</h2>\n<p dir=\"auto\">PMAT version 1.1.0 (2023.05.05)</p>\n<h2 tabindex=\"-1\" dir=\"auto\"><a id=\"user-content-author\" class=\"anchor\" aria-hidden=\"true\" href=\"#author\"><svg class=\"octicon octicon-link\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" height=\"16\" aria-hidden=\"true\"><path d=\"m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z\"></path></svg></a>Author</h2>\n<p dir=\"auto\">Changwei Bi <a href=\"mailto:bichwei@njfu.edu.cn\">bichwei@njfu.edu.cn</a><br>\nFuchuan Han <a href=\"mailto:hanfc@caf.ac.cn\">hanfc@caf.ac.cn</a></p>\n</article>","renderedFileInfo":null,"tabSize":8,"topBannersInfo":{"overridingGlobalFundingFile":false,"globalPreferredFundingPath":null,"repoOwner":"bichangwei","repoName":"PMAT","showInvalidCitationWarning":false,"citationHelpUrl":"https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-citation-files","showDependabotConfigurationBanner":false,"actionsOnboardingTip":null},"truncated":false,"viewable":true,"workflowRedirectUrl":null,"symbols":{"timedOut":false,"notAnalyzed":true,"symbols":[]}},"copilotAccessInfo":null,"csrf_tokens":{"/bichangwei/PMAT/branches":{"post":"hEdQ5NuWSy7_zHQrXmKr2AHhGNwq0B0CuvC_mfd-8FzcptzPjX9tjLdfBoc2dQDBivT96UB4jEgWuFUi1FYHnQ"}}},"title":"PMAT/README.md at main Â· bichangwei/PMAT","locale":"en"}